### PR TITLE
OR Query public API

### PR DESCRIPTION
**Pull request by MarkDuckworth**
Adding the OR Query public API

### CODE REVIEW

**Comment by MarkDuckworth** [comment]
@egilmorez please take a look at these documentation changes for the public API for the upcoming OR query feature.

**Review by egilmorez — COMMENTED** [review]
One high-level question before we pass the review to Mark A., thanks!

File: packages/firestore/test/integration/api/query.test.ts [hunk 1]
```diff
@@ -1543,34 +1827,29 @@ apiDescribe('Queries', (persistence: boolean) => {
...
+    it('supports order by equality', () => {
       const testDocs = {
         doc1: { a: 1, b: [0] },
         doc2: { b: [1] },
-        doc3: { a: 3, b: [2, 7] },
+        doc3: { a: 3, b: [2, 7], c: 10 },
         doc4: { a: 1, b: [3, 7] },
         doc5: { a: 1 },
-        doc6: { a: 2 }
+        doc6: { a: 2, c: 20 }
```
Comment (ehsannas) [inline packages/firestore/test/integration/api/query.test.ts:1838]: why has `c` been added to the test docs?

File: packages/firestore/test/integration/api/query.test.ts [hunk 2]
```diff
@@ -1525,14 +1817,6 @@ apiDescribe('Queries', (persistence: boolean) => {
...
 
       return withTestCollection(persistence, testDocs, async coll => {
-        // a==2 || b in [2,3]
-        await checkOnlineAndOfflineResultsMatch(
-          query(coll, or(where('a', '==', 2), where('b', 'in', [2, 3]))),
-          'doc3',
-          'doc4',
-          'doc6'
-        );
-
```
Comment (ehsannas) [inline packages/firestore/test/integration/api/query.test.ts:1535]: the name of the test can be updated now:
`can use or queries with not-in`

File: docs-devsite/firestore_lite.querycompositefilterconstraint.md [hunk 3]
```diff
@@ -0,0 +1,35 @@
+Project: /docs/reference/js/_project.yaml
```
Comment (egilmorez) [inline docs-devsite/firestore_lite.querycompositefilterconstraint.md:1]: Did I forget to press a button earlier today . . . ??

Anyway, my question was: why are these output files included in the PR? Generally we do those in a follow-on CL.  Thanks!

File: packages/firestore/test/integration/api/query.test.ts [hunk 4]
```diff
@@ -1464,6 +1450,318 @@ apiDescribe('Queries', (persistence: boolean) => {
...
 
+        // Test with limits without orderBy (the __name__ ordering is the tie breaker).
+        await checkOnlineAndOfflineResultsMatch(
+          query(coll, or(where('a', '==', 2), where('b', '==', 1)), limit(1)),
+          'doc2'
+        );
+      });
+    });
+
+    it('can use or queries with in and not-in', () => {
```
[the remark this answers is not in the document]
Comment (MarkDuckworth) [reply packages/firestore/test/integration/api/query.test.ts:1461]: fixed

File: packages/firestore/test/integration/api/query.test.ts [hunk 5]
```diff
@@ -1464,6 +1450,318 @@ apiDescribe('Queries', (persistence: boolean) => {
...
+        await checkOnlineAndOfflineResultsMatch(
+          query(coll, or(where('a', 'in', [2, 3]), where('b', 'in', [0, 2]))),
+          'doc1',
+          'doc3',
+          'doc6'
+        );
+
+        // Two IN operations on different fields with conjunction.
+        await checkOnlineAndOfflineResultsMatch(
+          query(coll, and(where('a', 'in', [2, 3]), where('b', 'in', [0, 2]))),
```
[the remark this answers is not in the document]
Comment (MarkDuckworth) [reply packages/firestore/test/integration/api/query.test.ts:1536]: Thanks. Removed

File: packages/firestore/test/integration/api/query.test.ts [hunk 6]
```diff
@@ -1464,6 +1450,318 @@ apiDescribe('Queries', (persistence: boolean) => {
...
+        // a IN [1,2,3] && a IN [0,1,4] should result in "a==1".
+        await checkOnlineAndOfflineResultsMatch(
+          query(
+            coll,
+            and(where('a', 'in', [1, 2, 3]), where('a', 'in', [0, 1, 4]))
+          ),
+          'doc1',
+          'doc4',
+          'doc5'
+        );
```
[the remark this answers is not in the document]
Comment (MarkDuckworth) [reply packages/firestore/test/integration/api/query.test.ts:1550]: Removed

File: packages/firestore/test/integration/api/query.test.ts [hunk 7]
```diff
@@ -1464,6 +1450,318 @@ apiDescribe('Queries', (persistence: boolean) => {
...
+
+        // a IN [2,3] && a IN [0,1,4] is never true and so the result should be an
+        // empty set.
+        await checkOnlineAndOfflineResultsMatch(
+          query(
+            coll,
+            and(where('a', 'in', [2, 3]), where('a', 'in', [0, 1, 4]))
+          )
+        );
```
[the remark this answers is not in the document]
Comment (MarkDuckworth) [reply packages/firestore/test/integration/api/query.test.ts:1559]: Removed

File: packages/firestore/test/integration/api/query.test.ts [hunk 8]
```diff
@@ -1464,6 +1450,318 @@ apiDescribe('Queries', (persistence: boolean) => {
...
+            and(
+              where('a', 'in', [1, 3]),
+              or(
+                where('a', 'in', [0, 2]),
+                and(where('b', '==', 2), where('a', 'in', [1, 3]))
+              )
+            )
+          ),
+          'doc3'
+        );
```
[the remark this answers is not in the document]
Comment (MarkDuckworth) [reply packages/firestore/test/integration/api/query.test.ts:1581]: Removed

File: packages/firestore/test/integration/api/query.test.ts [hunk 9]
```diff
@@ -1464,6 +1450,318 @@ apiDescribe('Queries', (persistence: boolean) => {
...
+            and(
+              where('b', 'in', [0, 3]),
+              or(
+                where('b', 'in', [1]),
+                and(where('b', 'in', [2, 3]), where('a', 'in', [1, 3]))
+              )
+            )
+          ),
+          'doc4'
+        );
```
[the remark this answers is not in the document]
Comment (MarkDuckworth) [reply packages/firestore/test/integration/api/query.test.ts:1596]: Removed

File: packages/firestore/test/integration/api/query.test.ts (hunk 2 above)
Comment (MarkDuckworth) [reply packages/firestore/test/integration/api/query.test.ts:1535]: Updated

File: packages/firestore/test/integration/api/query.test.ts (hunk 1 above)
Comment (MarkDuckworth) [reply packages/firestore/test/integration/api/query.test.ts:1838]: It was not added. `it('supports order by equality'` was not modified. The diff is invalid.

File: .changeset/sweet-rats-compete.md [hunk 10]
```diff
@@ -0,0 +1,5 @@
+---
+"@firebase/firestore": feat
```
Comment (dwyfrequency) [inline .changeset/sweet-rats-compete.md:2]: Update to 
```
'@firebase/firestore': minor
'firebase': minor
```

File: docs-devsite/firestore_lite.querycompositefilterconstraint.md (hunk 3 above)
Comment (markarndt) [reply docs-devsite/firestore_lite.querycompositefilterconstraint.md:1]: @egilmorez ... wrt to the feature, this content looks fine. Do you want resolution to this discussion before I give the single approval that will unblock?

File: .changeset/sweet-rats-compete.md (hunk 10 above)
Comment (MarkDuckworth) [reply .changeset/sweet-rats-compete.md]: Modified. Thank you for the review.

**Review by ehsannas — APPROVED** [review]

File: docs-devsite/firestore_lite.querycompositefilterconstraint.md (hunk 3 above)
Comment (egilmorez) [reply docs-devsite/firestore_lite.querycompositefilterconstraint.md:1]: Nope, go ahead Mark, thanks!  I'm curious see if a doc CL crosses my "desk."

**Review by egilmorez — APPROVED** [review]
iiuc, all the doc strings are in query.ts and reference.ts, both of which look fine to me.

**Review by dwyfrequency — APPROVED** [review]
LGTM

**Pull request merged**